### PR TITLE
Remove Sync instance for EitherT[Eval, Throwable, ?]

### LINKED ISSUE #170: Sync instance for EitherT[Eval, Throwable, ?]
As was mentioned in #169, currently there is a `Sync` instance for `EitherT[Eval, Throwable, ?]`. However this seems to be inconsistent, since it uses `Eval` to capture side-effects. (This is the same problem like in #78. Actually it's worse, since here it actually does use `Eval` in this way, instead of only suggesting it.)

### PR DESCRIPTION
**Pull request by durban**
This instance could be used to break referential transparency
in polymorphic code. Fixes #170.

### CODE REVIEW

**Comment by alexandru** [comment]
As expressed in #170, I'm against this change, as it goes against the goals of the `cats-effect` project, as was established.

**Comment by SystemFw** [comment]
@alexandru What do you think of https://github.com/typelevel/cats-effect/issues/170#issuecomment-376279711 though? That's quite the game changer, since you can break referential transparency using only perfectly valid code, similar to how Comonad + Sync does for naked Eval.

**Comment by johnynek** [comment]
Again, when we get opaque types all of this will be so nice. We can do these kinds of things without allocations. But with my suggestion we exchange the EitherT allocation for an EvalSync allocation and still keep a usable concrete Sync type around.

**Comment by LukaJCB** [comment]
We can probably get by with zero allocation using the `newtype` we're already using for `IO.Par`

**Comment by alexandru** [comment]
I don't have the bandwidth for a `newtype` proposal, I don't think that newtypes are a solution for Scala, but if that's the popular consensus then fine, plus I'm curious about the usability implications. Maybe I'm wrong and newtypes are usable.

@SystemFw 

> @alexandru What do you think of #170 (comment) though? That's quite the game changer, since you can break referential transparency using only perfectly valid code, similar to how Comonad + Sync does for naked Eval.

My opinion on that comment is that `Foldable` in Cats is broken because it has strict behavior and requires synchronous evaluation. That sample only works because `EitherT` and `Eval` both implement `Foldable`. Maybe we should question that decision instead.

That `Foldable` is broken a long held belief, ever since I noticed it can't be implement for streams capable of async computations or of suspending side effects, which isn't a surprise because it mirrors Haskell's `List`, but without the laziness, plus in spite of popular opinion, I don't find the `Foldable` type class to be very good or useful, being the replacement of `Iterable` that we need, but not what we deserve 😉

My argument still stands though — to remove an instance, a concrete law should be violated. Not going to fight against a pro decision in this regard. If you all feel that this should happen, then I'm fine with that.

Just a mention on https://github.com/typelevel/cats-effect/issues/170#issuecomment-376279711:

> Yes, I agree that if value would be called unsafeValue or similar, this problem would not exist. But it isn't called that ... And sure, this is subjective; but I find it immensely helpful that in Cats/cats-effect (and also in FS2) I can be sure that anything which breaks RT is prefixed with unsafe.

`Task` and `Coeval` in Monix do not use `unsafe` for describing `runAsync` and `run` respectively. I find that to be a misuse of the word `unsafe` (i.e. you use it, you lose it for actually unsafe, process imploding, money losing stuff) and I don't want this project to promote the notion that side effecting calls should be prefixed with "unsafe". It's a very anti-Scala thing to do. This isn't Haskell and that it happens for this `IO` implementation, such a decision is very context specific.

Haskell is an inspiration, but `cats-effect` is not a clone, as it is built for Scala and any decision has to be filtered through the lens of the ecosystem around it and its users.

---

TL;DR — if somebody can pick up the torch on the newtype proposal to see how that looks like, that would be great. And if you all agree that this instance should be removed, then I'll play along.

But it would be great to have that newtype proposal by 1.0.0-RC.

**Comment by alexandru** [comment]
The `Exec` PR at #178 is not done yet, but given the strong support for this change, I'd rather merge it now, rather than at `1.0.0`, because I'd like all breakage to happen in `1.0.0-RC`.
